src/credentials/StaticGroupDataProvider.cpp

```diff
@@ -0,0 +1,780 @@
+                {
+                    outGroup = entry.group;
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        void Release() override { mProvider.Release(this); }
+
+    private:
+        StaticGroupsProvider & mProvider;
+        Fabric * mFabric           = nullptr;
+        chip::EndpointId mEndpoint = 0;
+        uint16_t mIndex            = 0;
+    };
+
+    class FabricGroupStateIterator : public GroupStateIterator
+    {
+    public:
+        FabricGroupStateIterator(StaticGroupsProvider & provider, Fabric * fabric) : mProvider(provider), mFabric(fabric) {}
+
+        uint16_t Count() override
+        {
+            uint16_t count = 0;
+            for (uint16_t i = 0; this->mFabric && i < kStateEntriesMax; ++i)
+            {
+                const StateEntry & entry = this->mFabric->states[i];
+                if (entry.in_use)
+                {
+                    count++;
+                }
+            }
+            return count;
+        }
+
+        bool Next(GroupStateListEntry & outEntry) override
+        {
+            while ((this->mFabric != nullptr) && (this->mIndex < kStateEntriesMax))
+            {
+                const StateEntry & entry = this->mFabric->states[this->mIndex++];
+                if (entry.in_use)
+                {
+                    // Iterator has data available, copy the contents of the entry to the output
+                    outEntry.fabric_index  = mFabric->fabric_index;
+                    outEntry.list_index    = mOverallListIndex++;
+                    outEntry.group         = entry.group;
+                    outEntry.key_set_index = entry.key_set_index;
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        void Release() override { mProvider.Release(this); }
+
+    private:
+        StaticGroupsProvider & mProvider;
+        Fabric * mFabric           = nullptr;
+        uint16_t mIndex            = 0;
+        uint16_t mOverallListIndex = 0;
+    };
+
+    class AllGroupStateIterator : public GroupStateIterator
+    {
+    public:
+        AllGroupStateIterator(StaticGroupsProvider & provider) : mProvider(provider) {}
+
+        uint16_t Count() override
+        {
+            const Fabric * fabric = nullptr;
+
+            uint16_t count = 0;
+            for (chip::FabricIndex i = 0; i < kNumFabrics; ++i)
+            {
+                fabric = this->mProvider.GetExistingFabric(i);
+                for (uint16_t j = 0; fabric && j < kStateEntriesMax; ++j)
+                {
+                    const StateEntry & entry = fabric->states[j];
+                    if (entry.in_use)
+                    {
+                        count++;
+                    }
+                }
+            }
+
+            return count;
+        }
+
+        bool Next(GroupStateListEntry & outEntry) override
```

Comment:
Both this and FabricGroupStateIterator should likely use the `mGroupStateIndices` rather than direct structures, so that iteration is in list order. This will allow fulfilling the IM list ordering semantic promise, where entries of a full list are provided in inserted order, rather than another order. The current allocation scheme could allow out-of-order iteration

Reply:
I have reworked the group state list. Now there is a single list of group states that covers all fabrics, which better reflects the specification, and avoids the use of the intermediate index mapping.
To prevent any single fabric to use too many group state entries, a state count is kept per fabric.